### PR TITLE
Implement AWS KMS cryptography (includes #21)

### PR DESCRIPTION
**Pull request by puiterwijk**
This implements an abstracted signing crypto module that uses an Amazon
Web Services Key Management Service key to sign data.
It supports either KMS or local keys for verification.

Fixes: #5
Signed-off-by: Patrick Uiterwijk <patrick@puiterwijk.org>

This PR is based on top of #21 and #22.
It can be tested by setting the following environment variables:
- `AWS_ACCESS_KEY_ID`
- `AWS_SECRET_ACCESS_KEY`
- `AWS_DEFAULT_REGION`
- `AWS_KMS_TEST_KEY_ARN` - the ARN of the created test key
- `TEST_KEY_SIG_ALG` - a signature algorithm (`ES256`, `ES384` or `ES512`) that this key supports

This test currently does not yet run via GitHub CI, since there's a lack of secrets.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the Apache License Version 2.0, as specified in the LICENSE file of this repository.

### CODE REVIEW

**Comment by puiterwijk** [comment]
For openness: I personally only care about #21 and #22, but I decided to write and submit this one as well to ensure myself the API I designed in #21 allowed for implementing KMS.
I can update this PR if requested, but it's not as important to me if it's decided that this is not the right approach.

**Comment by puiterwijk** [comment]
Also, please do note: because the aws_sdk crates are not yet published to crates.io, this introduces a git dependency.
This will mean this cannot be pushed to crates.io once merged.
So perhaps this should go into a separate branch if you want this change, or make use of e.g. [Rusoto](https://www.rusoto.org) as AWS client.

**Comment by petreeftime** [comment]
Unfortunately this feature can't be enabled on crates.io yet, until the AWS SDK is pushed to crates.io, so I'll postpone merging this PR until that happens.

**Comment by puiterwijk** [comment]
@petreeftime cool. Do you want me to do a rebase again, or is this set to go in?

**Comment by petreeftime** [comment]
I wanted to also make the tests run with Github Actions before merging, but I need to create a role and a KMS key in our test account to make it work and didn't quite find time for it.

**Comment by puiterwijk** [comment]
Fair enough. Just let me know if you need me to do anything else.

**Comment by justin1121** [comment]
Any progress in getting this merged? Thanks!

**Comment by gal-fordefi** [comment]
Any news? We really need this for our deployment.

**Comment by petreeftime** [comment]
I've rebased the dev branch I had to the latest commit, but I still need to figure out the testing part for this feature.

**Comment by petreeftime** [comment]
I'll do some manual testing as soon as possible, merge and update the crate version to unblock you, and figure out the automated testing afterwards.

**Review by andraprs — DISMISSED** [review]

File: src/crypto/kms.rs
Comment (sabin-rapan) [inline src/crypto/kms.rs:37]: Creating a runtime in a library is super weird/anti-pattern.
Why not make your functions async and leave the runtime stuff to the application calling this library?

**Review by sabin-rapan — CHANGES_REQUESTED** [review]

File: Cargo.toml
Comment (meerd) [inline Cargo.toml:11]: If there is a need to do a version bump, I'd recommend aligning with the MSRV of the [aws-sdk-rust](https://github.com/awslabs/aws-sdk-rust), which is **1.58.1.**

File: .github/workflows/ci.yml
Comment (meerd) [inline .github/workflows/ci.yml:21]: Please see my comment below.

File: src/crypto/kms.rs
Comment (petreeftime) [reply src/crypto/kms.rs:37]: I need to take another look, but it required making the methods async, which meant it didn't fit with any of the other interfaces. Maybe there's a way around this, but I'm not sure what that is.

Comment (petreeftime) [reply src/crypto/kms.rs]: It's solved in the latest commit.

**Review by sabin-rapan — APPROVED** [review]

**Review by meerd — APPROVED** [review]

**Review by ddraga — APPROVED** [review]

**Pull request merged**
